### PR TITLE
Refactor to single settings file

### PR DESCRIPTION
**Pull request by JimacoMS4**
## Purpose

* ...Per cloud advocates refactor sample to use a single settings file.

## Does this introduce a breaking change?

```
[ ] Yes
[ x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code
Follow instructions in this topic: https://learn.microsoft.com/en-us/azure/developer/python/tutorial-deploy-python-web-app-azure-container-apps-02?tabs=azure-portal%2Ccreate-database-psql
```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

**Comment by JimacoMS4** [comment]
@pamelafox @kjaymiller @tonybaloney Can one of you please ack or provide feedback on these changes. Most have to do with collecting settings into a single settings.py file per 12-factor app methodology. We had an email thread about this a bit back. I'll resend with the PR to refresh your memory.

I tested this using an azure psff instance and used virtual environment, local container, and ACA hosted. IMHO the readme file could probably use more explicit instructions about running django admin commands i.e "python manage.py migrate" for each flavor (virtual env, local container, app service-hosted), but this sample is mainly to support running in ACA and the django admin instructions are covered in the associated mslearn topic. 

PS: Please freely suggest any edits you think are necessary. 

@bobtabor-msft FYI

File: azureproject/wsgi.py [hunk 1]
```diff
@@ -13,7 +13,8 @@
 
 # Check for the RUNNING_IN_PRODUCTION environment variable to see if we are running in Azure App Service
 # If so, then load the settings from production.py
-settings_module = 'azureproject.production' if 'RUNNING_IN_PRODUCTION' in os.environ else 'azureproject.settings'
+# settings_module = 'azureproject.production' if 'RUNNING_IN_PRODUCTION' in os.environ else 'azureproject.settings'
```
Comment (pamelafox) [inline azureproject/wsgi.py:16]: Remove commented line of code?

File: manage.py [hunk 2]
```diff
@@ -14,7 +14,8 @@ def main():
         load_dotenv('./.env')
 
     # When running on Azure App Service you should use the production settings.
-    settings_module = "azureproject.production" if 'RUNNING_IN_PRODUCTION' in os.environ else 'azureproject.development'
+    # settings_module = "azureproject.production" if 'RUNNING_IN_PRODUCTION' in os.environ else 'azureproject.development'
```
Comment (pamelafox) [inline manage.py:17]: Remove commented line of code. Oh and comment above!

File: azureproject/wsgi.py (hunk 1 above)
Comment (pamelafox) [reply azureproject/wsgi.py:16]: Also remove comments above

File: azureproject/settings.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 # SECURITY WARNING: keep the secret key used in production secret!
 # Use this py command to create secret 
 # python -c 'import secrets; print(secrets.token_hex())'
-SECRET_KEY = os.getenv('SECRET_KEY')
+# SECRET_KEY = os.getenv('SECRET_KEY')
```
Comment (pamelafox) [inline azureproject/settings.py:10]: Why is this commented out?

File: azureproject/settings.py [hunk 4]
```diff
@@ -93,3 +93,67 @@
 # https://docs.djangoproject.com/en/4.0/ref/settings/#default-auto-field
 
 DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'
+
+if 'RUNNING_IN_PRODUCTION' in os.environ:
+    # Configure allowed host names that can be served and trusted origins for Azure Container Apps.
+    ALLOWED_HOSTS = ['.azurecontainerapps.io']
+    CSRF_TRUSTED_ORIGINS = ['https://*.azurecontainerapps.io']
+    DEBUG = False
+    DEBUG_PROPAGATE_EXCEPTIONS = True
```
Comment (pamelafox) [inline azureproject/settings.py:102]: Why is this True?

https://docs.djangoproject.com/en/5.0/ref/settings/#debug-propagate-exceptions

I think it's typically False in production?

File: azureproject/settings.py [hunk 5]
```diff
@@ -93,3 +93,67 @@
...
+            'PASSWORD': os.environ['AZURE_POSTGRESQL_PASSWORD'], 
+        }
+    }
+else:
+    # SECURITY WARNING: don't run with debug turned on in production!
+    DEBUG = True
+
+    # Use this py command to create secret 
+    # python -c 'import secrets; print(secrets.token_hex())'
+    SECRET_KEY = os.getenv('LOCAL_SECRET_KEY')
```
Comment (pamelafox) [inline azureproject/settings.py:129]: I guess the idea here is to be doubly sure they don't reuse a local secret_key in production? That's fine, but in that case, remove the code at the top.

File: azureproject/settings.py [hunk 6]
```diff
@@ -93,3 +93,67 @@
...
+
+    # Don't use Whitenoise to avoid having to run collectstatic first.
+    STATICFILES_STORAGE = 'django.contrib.staticfiles.storage.StaticFilesStorage'
+
+    ALLOWED_HOSTS = ['*']
```
Comment (pamelafox) [inline azureproject/settings.py:134]: Is this needed? DEBUG=True has special handling for ALLOWED_HOSTS: https://docs.djangoproject.com/en/5.0/ref/settings/#allowed-hosts

File: azureproject/settings.py [hunk 7]
```diff
@@ -93,3 +93,67 @@
...
+
+    # Configure connection setting for PostgreSQL instance.
+    # Set these environment variables in the .env file for this project.  
+    if 'USE_REMOTE_POSTGRESQL' in os.environ:
```
Comment (pamelafox) [inline azureproject/settings.py:138]: Eek, are customers asking for this? I stay far away from prod databases when I'm local, unless I have a good way to explicitly turn on read-only access to them.

File: azureproject/settings.py [hunk 8]
```diff
@@ -93,3 +93,67 @@
...
+    # Configure database connection for Azure PostgreSQL Flexible server instance.
+    # AZURE_POSTGRESQL_HOST is the full URL.
+    # AZURE_POSTGRESQL_USERNAME is just name without @server-name.
+    DATABASES = {
+        'default': {
+            'ENGINE': 'django.db.backends.postgresql',
+            'NAME': os.environ['AZURE_POSTGRESQL_DATABASE'],
+            'HOST': os.environ['AZURE_POSTGRESQL_HOST'],
+            'USER': os.environ['AZURE_POSTGRESQL_USERNAME'],
+            'PASSWORD': os.environ['AZURE_POSTGRESQL_PASSWORD'],
```
Comment (pamelafox) [inline azureproject/settings.py:120]: You may want to specify sslmode=required. We often find that's necessary when connecting to Azure Postgres databases. However, the Postgres add-on sets a POSTGRES_SSL env variable to indicate whether it should be on/off. This is how we deal with it:
https://github.com/Azure-Samples/azure-django-postgres-flexible-aca/blob/main/src/project/settings.py#L107C1-L108C39

And then we set it to require in the infra:
https://github.com/Azure-Samples/azure-django-postgres-flexible-aca/blob/5d30fcab0354c14453575a8ee0ba29c0dadc87e7/infra/web.bicep#L65

You may not need that complexity depending on what postgres options you're supporting.

File: azureproject/settings.py [hunk 9]
```diff
@@ -93,3 +93,67 @@
...
+    CSRF_TRUSTED_ORIGINS = ['https://*.azurecontainerapps.io']
+    DEBUG = False
+    DEBUG_PROPAGATE_EXCEPTIONS = True
+
+    # SECURITY WARNING: keep the secret key used in production secret!
+    # Use this py command to create secret 
+    # python -c 'import secrets; print(secrets.token_hex())'
+    SECRET_KEY = os.getenv('AZURE_SECRET_KEY')
+
+    STATICFILES_STORAGE = 'whitenoise.storage.CompressedStaticFilesStorage'
```
Comment (pamelafox) [inline azureproject/settings.py:109]: I think I now use whitenoise locally as well in my django apps for consistency, but I do not feel strongly about it.

**Review by pamelafox — CHANGES_REQUESTED** [review]
See my comments

File: azureproject/wsgi.py (hunk 1 above)
Comment (JimacoMS4) [reply azureproject/wsgi.py:16]: Tx fixing on next push

File: manage.py (hunk 2 above)
Comment (JimacoMS4) [reply manage.py:17]: Tx fix on next push

File: azureproject/settings.py (hunk 4 above)
Comment (JimacoMS4) [reply azureproject/settings.py:102]: Truthfully -- I have no idea. It was set that way in production.py in the main branch. You're right, it certainly doesn't appear that it should ever be set in production. I did see a comment in a repo that Jay linked me to that said it is set when DEBUG is enabled (not production) to enable the VS Code debugger to break on raised exceptions. Just copied the line from that repo in the next push.

File: azureproject/settings.py (hunk 5 above)
Comment (JimacoMS4) [reply azureproject/settings.py:129]: Updated the SECRET_KEY stuff to be similar to the repo that you and Jay have pointed me to on next push.

File: azureproject/settings.py (hunk 6 above)
Comment (JimacoMS4) [reply azureproject/settings.py:134]: Thanks, it sure doesn't look like it. :). Just set it to [] on next push.

File: azureproject/settings.py (hunk 7 above)
Comment (JimacoMS4) [reply azureproject/settings.py:138]: @pamelafox isn't this a valid debugging scenario? i.e. running locally against a DB hosted in Azure to flush out any bugs that might arise due to implementation differences between the cloud and local implementation of the DB? It doesn't necessarily have to be the production DB.

**Comment by bobtabor-msft** [comment]
@JimacoMS4  What is the status of this change? Can I merge it or is there work left to do?

**Comment by JimacoMS4** [comment]
@bobtabor-msft -- sorry, missed your previous comment. Turns out this would require changes to the Flask tutorial as well to fit into our docs. I'm closing it and deleting the branch.

**Comment by JimacoMS4** [comment]
Closing -- past its sell by date :)

**Pull request closed without merging**